### DISCUSSION TITLE
In-memory JWT setup

### DISCUSSION THREAD

**Discussion by pattruong** [urql-graphql/urql#1178] [Q&A] [answered]
Hello!
I've been using urql for a few months now and slowly been building my auth layer using the guide in the [doc](https://formidable.com/open-source/urql/docs/advanced/authentication/) along with referencing any GitHub issues I found. I noticed every example using token storage on `localStorage`. I initially wanted to avoid that due to various people suggesting otherwise (including [howtographql tutorial](https://www.howtographql.com/react-urql/5-authentication/) with urql). I was wondering if anyone had an example of storing a JWT in React state. 

My goal is to follow this architecture ([Hasura ref](https://hasura.io/blog/best-practices-of-using-jwt-with-graphql/)), which I would summarize as:
- Get a JWT and store in state
- Get a refresh token which is stored in `httpOnly` cookie
- Silent refresh to get new JWT on every session initialization (new tab or refresh site)

My current issue is that I'm using React state to keep track of my JWT and it appears that my urql client is being initialized twice. This makes me believe that the re-renders due to the state are re-initializing my client, triggering my `getAuth` function to run twice in quick succession. This ends up calling my `/refresh-token` endpoint twice for every refresh, not just initialization.

My `App.js`:
```javascript
function App() {
  const history = useHistory();
  const { pathname } = useLocation();

  const [token, setToken] = useState(null);
  const [tokenExpiry, setTokenExpiry] = useState(null);
  const [hasInit, setHasInit] = useState(localStorage.getItem('hasLoggedIn') !== 'yes')

  const client = useMemo(() => {
    return createClient({
      url: GRAPHQL_API,
      exchanges: [
        devtoolsExchange,
        dedupExchange,
        cacheExchange,
        authExchange({
          getAuth: async ({ authState }) => {
            if (!authState) {
              if (token) {
                console.log("HERE (1) - token:", token.slice(-6));
                return { token };
              }
              if (!hasInit) {
                console.log("HERE (2) - INIT:", hasInit);
                const { ok, ...data } = await getTokenFromRefresh();
                console.log('REFRESH CALLED. TOKEN ->', data.token.slice(-6))

                if (ok) {
                  setToken(data.token);
                  setTokenExpiry(data.tokenExpiry);
                  setHasInit(true);
                  return { token: data.token }
                }
              }
              return null;
            }
            const { ok, ...data } = await getTokenFromRefresh();

            if (ok) {
              setToken(data.token);
              setTokenExpiry(data.tokenExpiry);
              return { token: data.token }
            }

            // --- LOGOUT BLOCK ---
            const isAuthOnlyPath = authOnlyPaths.find(({ path, exact, strict }) => matchPath(pathname, {
              path,
              exact,
              strict
            }));

            // remove refresh token cookie
            await axios.post(AUTH_API + '/logout', { withCredentials: true });

            // to support logging out from all windows
            window.localStorage.setItem('logout', Date.now());
            window.localStorage.clear();
            setToken(null);
            setTokenExpiry(null);

            if (isAuthOnlyPath) {
              history.replace('/login');
            }

            return null;
          },
          addAuthToOperation,
          didAuthError: ({ error }) => {
            const hasError = error.graphQLErrors.some(e => e.extensions?.code === 'invalid-jwt');
            console.log("HERE (4) - auth error:", error.graphQLErrors, hasError);
            return hasError;
          },
        }),
        fetchExchange,
      ],
    })
  }, []);

  return (
    <UrqlProvider value={client}>
      <UserProvider authOnlyPaths={authOnlyPaths} setToken={setToken} setTokenExpiry={setTokenExpiry} token={token}>
        <Switch>
          {/* Routes here */}
        </Switch>
      </UserProvider>
    </UrqlProvider>
  );
}

export default App;
```

Note I'm trying to use `useMemo` on my client in hopes of reducing re-initializations between re-render. It hasn't solved the issue, but I found that the re-initializations can happen to 5 times without it. I assume that has to do with the `setState`(ex. `setToken`, `setHasInit`) calls.

![Screen Shot 2020-11-24 at 10 21 14 AM](https://user-images.githubusercontent.com/6396042/100135651-cdc4aa00-2e3e-11eb-8e37-90166fabc74a.png)
- You can see here how the refresh initialization gets called twice and despite `await`, it seems like it's just executing concurrently.

I've almost given up on this setup and have another implementation using the traditional `localStorage` setup, but I think it would be useful, if it is possible, to have an example of using urql auth with an in-memory token storage solution.

Thanks for any ideas and help 😃

**Comment by JoviDeCroock** [comment, marked as answer]
The reason your client is created twice would be [`StrictMode`](https://reactjs.org/docs/strict-mode.html), the token-refresh gets called twice because you are calling it eagerly in your `getAuth` function, it seems better to only refresh your token when an error occurs or when you're sure that the token has expired.

**Reply by pattruong** [reply]
I never knew that StrictMode did that until now. That makes a lot more sense.

> the token-refresh gets called twice because you are calling it eagerly in your getAuth function

By eagerly,  are you referring to how I'm using the `hasInit` check to try to call a refresh token on page load? I assume since `getAuth` is called on client initialization and StrictMode also initializes it twice, that this is now what's causing the refresh call to double. Is that a correct understanding?

I'm not sure if this is straightforward based on how streams work, but I noticed sometimes in the past that after initialization, that querying will trigger `getAuth` twice (checked with a log). Is that to be expected? I know it triggers it before running the `fetchExchange`, but does it also trigger it on the Operation Result? Something like this:
```
Operation (query) -> ...exchanges -> authExchange (getAuth called here) -> fetchExchange -> 

Client executes query

results -> fetchExchange -> authExchange (2nd getAuth call?) -> ...exchanges
```
The `getAuth` being triggered twice surprised me, because my understanding is that `getAuth` only runs on query/mutation operation initialization and if `didAuthError` returns `true`.

Referencing these lines from https://github.com/FormidableLabs/urql/pull/939:
> Once the exchange is created, we trigger getAuth for the first time.

> Once the request has been done, if an error occurs, then `didAuthError` is called to check whether the error was an auth error. If this returns true, then the auth library will call `getAuth`


Thanks for the help

**Reply by pattruong** [reply]
@JoviDeCroock I just tried my query again without wrapping `StrictMode` around my app and now it doesn't double call `getAuth` on each query. Bit surprised that `StrictMode` also triggers a double call for exchanges and not just double initialization of the client.

**Reply by kitten** [reply]
@pattruong Double render means that your `useMemo` is instantiated twice, which instantiates the `urql` Client twice, which initialises by calling `getAuth` twice. So it's a natural effect of managing this as state in React basically.

As you already identified you could also manage your auth state separately, for instance in an event hub. But since that requires more time there's another alternative.

`getAuth` is ultimately eagerly called, but `willAuthError` is lazily called. So if you _don't_ eagerly refresh your auth state in `getState` and instead just return a placeholder on initialisation if you don't have any auth state, you can use `willAuthState` to then re-trigger the refresh when an operation is sent.

This is different than what you currently have since it's not necessary to refresh the auth state immediately on initialisation.

The alternative of creating separate instances for managing your auth state is interesting as well though. You could keep the auth state in global singletons (or instances that you create separately) and then pass that to React and pass it to `urql` (where you may instantiate the client outside of a React component) separately to isolate that state. That being said, with `StrictMode` you'll always have to handle double renders. It's an approximation of making sure you're making all your effects and components double-render-safe.

**Reply by pattruong** [reply]
@kitten that's good info on the double instantiations.

I haven't seen `getState` and `willAuthState` before in the docs, can you point me to any docs or PRs on those config functions?

**Reply by kitten** [reply]
Oh sorry, those are typos. I meant `getAuth` instead of `getState`, and `willAuthError` rather than `willAuthState` https://formidable.com/open-source/urql/docs/advanced/authentication/#configuring-willautherror

Generally as you can see in the documentation `return null;` is a common initialisation in there as `willAuthError` doesn't have to care about why some auth state is invalid during initialisation. That's because it may be used as an initialiser and on an auth error (for a refresh, etc) at the same time.
